For large Layout: Calculate height with row height,size and gap

MrBearPresident opened the pull request:
Hey @Clooos here my first pull request. 
Haven't gone through all the code yet but this is a start. 

![image](https://github.com/user-attachments/assets/08ae14db-3892-415d-a7b4-067294741455)

Clooos commented:
Hi! I'm so glad that you made you first PR! 😄

I just have some suggestions, because it was something I wanted to do already, it would be great if the icon/name/state stay fixed at the top, and the sub-buttons fixed at the bottom.

This would better to enjoy the additional space that this offer.

What do you think? And don't hesitate if you have any suggestions 🙂

MrBearPresident commented:
Already got it working on the button-cards. 
![image](https://github.com/user-attachments/assets/79324bd2-3127-4521-ac4b-b53beb52fbb9)

The other cards will follow. I will not do the: empty-column, separator, button-stack, pop-up as I don't see a lot of usage for this.

Clooos commented:
That's amazing! Thanks a lot! 😀

Clooos commented:
Hi again! I should be able to try that tomorrow, can't wait!

MrBearPresident commented:
Not ready yet.  Will have some more work. So no need to try for now.

Clooos commented:
Ok good to know, tell me when I will be able to try it 🙂

And thanks again!

MrBearPresident commented:
![image](https://github.com/user-attachments/assets/5386b13f-94b3-492f-ba0c-d9ab95b5a4ad)

![image](https://github.com/user-attachments/assets/a68b5962-171b-4b25-87da-d977556a2061)

![image](https://github.com/user-attachments/assets/4ae1bdec-710b-4527-b952-dc4997a22c1a)


Possibly some fine-tuning still to do, but this can already be evaluated.

MrBearPresident commented:
You'll notice some extra commented code for the sub-buttons. 
What i wanted to do was the following.

I always worked with the following grid:
| i | n | c | a |
|:-:|:-:|:-:|:-:|
| b | b | b | b |

When there was only 1 row
- icon in 'i'
- name in 'n'
- sub-buttons in 'c'
- all other features in 'a'

when there were multiple rows
-sub-buttons in 'b' but when overflowing also in 'c'

but this was very unrelliable.

Clooos commented:
Good work again! Even if I think that it should be improved a bit more, even if I'm still unsure how exactly. What's bugging me is that the space behind is not used fully when in a 2 rows config.

Maybe something like this?

<img width="973" alt="image" src="https://github.com/user-attachments/assets/739919be-508c-41b6-b6ed-3e34cf0a821a">

And in a `card_layout: large-2-row` the sub-buttons could take more space to be less condensed.

What do you think? I hope I'm not asking too much.

Clooos commented:
We should also handle when the card is not large enough, and I understand that this is not that easy 😅

<img width="557" alt="image" src="https://github.com/user-attachments/assets/71555442-e1b4-42f8-ae65-44885b1a5109">

MrBearPresident commented:
How should it resolve such a thing?
I could not think of a good solution.

MrBearPresident commented:
> Good work again! Even if I think that it should be improved a bit more, even if I'm still unsure how exactly. What's bugging me is that the space below is not used fully when in a 2 rows config.
> 
> Maybe something like this?
> 
> <img alt="image" width="973" src="https://private-user-images.githubusercontent.com/36499953/393436564-739919be-508c-41b6-b6ed-3e34cf0a821a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2MDA5NzksIm5iZiI6MTczMzYwMDY3OSwicGF0aCI6Ii8zNjQ5OTk1My8zOTM0MzY1NjQtNzM5OTE5YmUtNTA4Yy00MWI2LWI2ZWQtM2UzNGNmMGE4MjFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDE5NDQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyZmVkODI4ODJlNGRmYzRkNmZjOGE0OWM4OGY2NTU2MjU3MDRmOGY1NTgzNjMyY2ZmZTljM2I2MDNjNjZiYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5kSQgttEwYuFp9FTzlTBBXs43d76a0LP0lhLIsEMaLY">
> And in a `card_layout: large-2-row` the sub-buttons could take more space to be less condensed.
> 
> What do you think? I hope I'm not asking too much.

Yes indeed, always aligning center is better.
About the spacing I would like to go with the second one. 
If people want the first they could adjust that through the styles sections. 

The third design would be easy doable.

Clooos commented:
I guess you're right, I'm probably overthinking.

This will be a major new possibility for Bubble Card, I will try it when I got some time.

MrBearPresident commented:
I think it is ready. 

But we have to discus maybe the following 'issue'.
The layout of the cards doesn't 100% line up. 

![image](https://github.com/user-attachments/assets/69841960-248d-4eeb-a99a-38467294cb63)

Top = Large layout
Bottom = Original layout

Clooos commented:
This layout was made to support the new section view, the more I think of it and the more I think we should remove the original layout, it was a bit too compact anyway. What do you think?

MrBearPresident commented:
I would keep the original layout in, but make the "large" the default one. 
But I would only do that once all the bugs are out.

MrBearPresident commented:
Then it will need further adjustment because, when the card is not in a section view it does not work.

Clooos commented:
It does work outside of a section view, I mean all the large layouts does, what do you mean?

MrBearPresident commented:
I solved It already so no problem anymore.

Clooos commented:
Hi again! I had some time to try your fork. Here's my first feedback 🙂

First, here's my dashboard with your PR:

<img width="627" alt="image" src="https://github.com/user-attachments/assets/a29e3c4e-c2b2-470f-87e4-f029d65eee22">

And here's what it looked like before:

<img width="621" alt="image" src="https://github.com/user-attachments/assets/d4d37931-15eb-4fa1-bb68-b1ff1e379085">

The icons are not round with the large layout (but not in all card types):

<img width="131" alt="image" src="https://github.com/user-attachments/assets/7da11490-536a-452a-a807-c32c2f7dba39">

The border radius is different when there is 2 rows:

<img width="84" alt="image" src="https://github.com/user-attachments/assets/22406e88-9994-4622-b963-0bcd4178f9d8"><img width="79" alt="image" src="https://github.com/user-attachments/assets/dd72a6d1-20fe-4488-96a3-e2d1fd001e66">

When there is 2 rows all buttons should be below (and not just the sub-buttons):

<img width="533" alt="image" src="https://github.com/user-attachments/assets/5b982e66-19c0-4177-b6de-ff629060d4e8">

The icon of the button switch is covered by the background color (after/before):

<img width="146" alt="image" src="https://github.com/user-attachments/assets/ee45e171-5a36-4182-b97c-b6cdb5274512"><img width="93" alt="image" src="https://github.com/user-attachments/assets/a2582feb-18bb-4c81-98ac-12632f16e96c">


There are maybe more issues, but beside that, that's a good start!

Thanks again!

Clooos commented:
I disagree, because the main use case for this is to have the possibility to display two or more cards on a same line, this way it will be possible to display two media players/climate cards next to each other on mobile. To me there is no sense to only display the sub-buttons below, because default buttons looks exactly the same for the end user.

MrBearPresident commented:
I fix the small issues you mentioned. 
I don't seem to have the problem of the background color.
![image](https://github.com/user-attachments/assets/38db7909-2a06-4b13-bc86-fd60b9bf05a1)

For the layout, I will try to add a toggle so people can choose?
For the layout you propose the only workable outcome I can think of is something like this: 
![image](https://github.com/user-attachments/assets/ab12e9a4-2232-4ba4-87b5-1e04eb556710)

Clooos commented:
I will try to test that today! 😀

For your example that's perfect, but the more empty line should be below (the one with only one sub-button).

Thanks again!

MrBearPresident commented:
Alternative layout is possible for media-player and climate card. 
Cover card is truly impossible unless we do a complete overhaul of the cover card.
For the select-card I think it is best the arrow stays where it is now.
Button didn't need work.

MrBearPresident commented:
![Screenshot_20241212_041334_Vivaldi](https://github.com/user-attachments/assets/822fcc52-96e7-4b9a-8c34-4c4c66f29e9a)
![Screenshot_20241212_041621_Vivaldi](https://github.com/user-attachments/assets/bd49ab4a-6bc6-4a53-8aaa-7f22fefdbe01)

MrBearPresident commented:
4 problems I've see already
- auto centering doesn't work anymore (when there are no sub-buttons) 
- in climate card, temp control needs more clearance on the right side
- select-sub-buttons open badly in the alternatieve layout

MrBearPresident commented:
Should be all good now...

Clooos commented:
Nice progress! I haven't tried it yet but I see on your screenshots that the sub-buttons are not perfectly centered when they are below, the margin on the left is shorter than the one on the right.

MrBearPresident commented:
> but I see on your screenshots that the sub-buttons are not perfectly centered when they are below, the margin on the left is shorter than the one on the right.

This is also fixed, but did not mention this in my commit.
![image](https://github.com/user-attachments/assets/603ed964-1c0c-4bcc-82cf-4d96cd9949da)

> conditional logic for the pop-up triggers
Nice!
I wasn't planning on working on new features until this one is merged. After that I want to continue organizing and handing  some issues.

Clooos commented:
Nice! I will have some time today to try this, I'll let you now my feedback. I will also try to fix the remaining important issues for v2.3.1.

MrBearPresident commented:
> I disagree, because the main use case for this is to have the possibility to display two or more cards on a same line, ...

So this was what I was thinking about: 
![image](https://github.com/user-attachments/assets/a429104e-b9cc-4601-ae3e-6689b9c7d829)

- Upper one is pure ChromeCasf entity
  - When somebody Cast music this one is needed, music assistant doesn't always show the info. 
  - Added sub-buttons are the direct control of the JBL-soundbar. 

-  Lower one is music assistant entity
  - Need this for when playing through this the next and forward buttons don't work on the standard ChromeCast entity.
  - Added sub-buttons are playlist that I can directly start with one-click.

Clooos commented:
That's nicer indeed when the card is taking the full column width, but I'm glad you added the possibility to put these below for the shorter column cases.

Looks great, thank you again!

MrBearPresident commented:
For people that want to test: [bubble-card.zip](https://github.com/user-attachments/files/18255455/bubble-card.zip)

To use unzip file above and follow the ["without HACS"-method](https://github.com/Clooos/Bubble-Card/tree/main?tab=readme-ov-file#installation).

**PLEASE do not make new issues about this PR** If you find problems, (and they are not present on the main branch) please report them in this thread. 


Features:
- All cards should work the same
- Large layout has been changed this you will only see when you give it 2 rows. 
  - It changed a lot so some of your custom layouts may not work anymore. 
  - Media-cards and Climate cards can switch between 2 different designs. (see picture below)
    - option1
![Screenshot 2024-12-26 221359](https://github.com/user-attachments/assets/05b985ad-a388-412f-a9d7-8501f6e9b3bc)
    - option 2    
![image](https://github.com/user-attachments/assets/88084441-bf96-40ca-9b21-df4b159217d2)

HyperCriSiS commented:
Do I oversee something in the new beta or is it not merged yet? 😊

MrBearPresident commented:
> Do I oversee something in the new beta or is it not merged yet? 😊

This exact PR wont be merged.  This was to complex to maintain because it was in al css files and in the subbutton creation, and... With every PR that got merged or direct push from @Clooos I had to refractor lots of code.

@Clooos was also planning to restructure the cards. I chose to close this one and make a new one.

- #1172

This doesn't have the nice feature of putting the subbuttons below.

HyperCriSiS commented:
Is there any way to accomplish this in the new version with CSS?

MrBearPresident commented:
I can probably make this as a module once #1171 is merged by @Clooos.

ddkedr commented:
> I can probably make this as a module once #1172 is merged by @Clooos.

That would be great! Really waiting for that! If you need any assistance in testing this, let us know ✊
